internal/pool/pool_single.go

```diff
@@ -37,6 +37,9 @@ func (p *SingleConnPool) Put(ctx context.Context, cn *Conn) {}
 func (p *SingleConnPool) Remove(ctx context.Context, cn *Conn, reason error) {
 	p.cn = nil
 	p.stickyErr = reason
+	if reason == nil && ctx != nil {
+		p.stickyErr = ctx.Err()
+	}
```

Comment:
Just to make sure I'm understanding it correctly.

When a query/connection is canceled, we will not receive an error as a reason. Hence the nil check here. So when we are removing it from the connection pool and we haven't been given a reason; check to see if there is an error on the context itself, which will have either a cancel error or some kind of deadline exceed.

Is that correct?

---

If so could that be left as a hand wavy comment here? The diagram in the PR is excellent for the flow of events, but hard to back reference that if changes ever need to be made in the future. So having comments is always nice.

Reply:
@elliotcourant , yes, that is correct.
The `reason` (`err`) is nil, since there is no error returns from the query, instead ctx.Done() returns with `context cancelled`.
Since `err` is nil, so it did not return an error in line 76.
https://github.com/go-pg/pg/blob/f7e1c98fed3044ae4e195a0e063f69185fb1885b/base.go#L76
And in line 80, then `nil` conn object check its `Inited` field and caused an NPE:
https://github.com/go-pg/pg/blob/f7e1c98fed3044ae4e195a0e063f69185fb1885b/base.go#LL80C1-L80C1
<img width="888" alt="image" src="https://github.com/go-pg/pg/assets/107882527/2f9f870a-155a-448d-a7ba-1148445778e8">

Good point! I will add a comment for explaining the logic.

Reply:
Added a comment: 2755f6daad1fb5d0eaaa204c48afc7e4ba63a1c4